File: frontend/src/services/slack.ts
import api from './api'

// Types
export interface SlackConnectionStatus {
  connected: boolean
  team_id?: string
  team_name?: string
  bot_user_id?: string
}

export interface SlackChannel {
  id: string
  name: string
  is_private: boolean
  is_member: boolean
  num_members?: number
}

export type StorageMode = 'FULL_CONTENT' | 'METADATA_ONLY' | 'EMBEDDINGS_ONLY'

export interface SlackWorkspaceConfig {
  id: number
  organization_id: string
  team_id: string
  allowed_channel_ids: string[]
  storage_mode: StorageMode
  default_agent_id?: string
  created_at: string
  updated_at: string
}

export interface AgentSlackConfig {
  id: number
  organization_id: string
  team_id: string
  agent_id: string
  channel_id: string
  channel_name: string
  enabled: boolean
  respond_to_mentions: boolean
  respond_to_reactions: boolean
  respond_to_commands: boolean
  reaction_emoji: string
  created_at: string
  updated_at: string
}

export interface AgentSlackConfigCreate {
  channel_id: string
  channel_name: string
  enabled?: boolean
  respond_to_mentions?: boolean
  respond_to_reactions?: boolean
  respond_to_commands?: boolean
  reaction_emoji?: string
}

export interface AgentSlackConfigUpdate {
  enabled?: boolean
  respond_to_mentions?: boolean
  respond_to_reactions?: boolean
  respond_to_commands?: boolean
  reaction_emoji?: string
}

export interface SlackWorkspaceConfigUpdate {
  allowed_channel_ids?: string[]
  storage_mode?: StorageMode
  default_agent_id?: string
}

/**
 * Check if Slack is connected for the current organization
 */
export const checkSlackConnection = async (): Promise<SlackConnectionStatus> => {
  try {
    const response = await api.get('/slack/status')
    return response.data
  } catch (error) {
    console.error('Error checking Slack connection:', error)
    return { connected: false }
  }
}

/**
 * Get the Slack authorization URL
 */
export const getSlackAuthUrl = () => {
  return `${import.meta.env.VITE_API_URL}/slack/authorize`
}

/**
 * Disconnect Slack integration
 */
export const disconnectSlack = async () => {
  try {
    const response = await api.delete('/slack/disconnect')
    return response.data
  } catch (error) {
    console.error('Error disconnecting Slack:', error)
    throw error
  }
}

/**
 * Get Slack channels the bot is a member of
 */
export const getSlackChannels = async (): Promise<SlackChannel[]> => {
  try {
    const response = await api.get('/slack/channels')
    return response.data
  } catch (error) {
    console.error('Error getting Slack channels:', error)
    throw error
  }
}

/**
 * Get workspace configuration
 */
export const getSlackWorkspaceConfig = async (): Promise<SlackWorkspaceConfig> => {
  try {
    const response = await api.get('/slack/workspace-config')
    return response.data
  } catch (error) {
    console.error('Error getting Slack workspace config:', error)
    throw error
  }
}

/**
 * Update workspace configuration
 */
export const updateSlackWorkspaceConfig = async (
  config: SlackWorkspaceConfigUpdate
): Promise<SlackWorkspaceConfig> => {
  try {
    const response = await api.put('/slack/workspace-config', config)
    return response.data
  } catch (error) {
    console.error('Error updating Slack workspace config:', error)
    throw error
  }
}

/**
 * Get agent's Slack channel configurations
 */
export const getAgentSlackConfig = async (agentId: string): Promise<AgentSlackConfig[]> => {
  try {
    const response = await api.get(`/slack/agent-config/${agentId}`)
    return response.data
  } catch (error) {
    console.error('Error getting agent Slack config:', error)
    return []
  }
}

/**
 * Create a new agent-to-channel configuration
 */
export const createAgentSlackConfig = async (
  agentId: string,
  config: AgentSlackConfigCreate
): Promise<AgentSlackConfig> => {
  try {
    const response = await api.post(`/slack/agent-config/${agentId}`, config)
    return response.data
  } catch (error) {
    console.error('Error creating agent Slack config:', error)
    throw error
  }
}

/**
 * Update an agent's Slack channel configuration
 */
export const updateAgentSlackConfig = async (
  agentId: string,
  configId: number,
  config: AgentSlackConfigUpdate
): Promise<AgentSlackConfig> => {
  try {
    const response = await api.put(`/slack/agent-config/${agentId}/${configId}`, config)
    return response.data
  } catch (error) {
    console.error('Error updating agent Slack config:', error)
    throw error
  }
}

/**
 * Delete an agent's Slack channel configuration
 */
export const deleteAgentSlackConfig = async (
  agentId: string,
  channelId: string
): Promise<void> => {
  try {
    await api.delete(`/slack/agent-config/${agentId}/${channelId}`)
  } catch (error) {
    console.error('Error deleting agent Slack config:', error)
    throw error
  }
}

File: backend/app/models/schemas/slack.py
"""
ChatterMate - Slack Schemas
Copyright (C) 2024 ChatterMate

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from uuid import UUID
from datetime import datetime
from enum import Enum


class StorageModeEnum(str, Enum):
    """Storage mode for Slack messages."""
    FULL_CONTENT = "FULL_CONTENT"
    METADATA_ONLY = "METADATA_ONLY"
    EMBEDDINGS_ONLY = "EMBEDDINGS_ONLY"


# ==================== Connection Status ====================

class SlackConnectionStatus(BaseModel):
    """Response schema for Slack connection status."""
    connected: bool
    team_id: Optional[str] = None
    team_name: Optional[str] = None
    bot_user_id: Optional[str] = None


# ==================== Channel Schemas ====================

class SlackChannelResponse(BaseModel):
    """Schema for a Slack channel."""
    id: str
    name: str
    is_private: bool = False
    is_member: bool = False
    num_members: Optional[int] = None


# ==================== Workspace Config Schemas ====================

class SlackWorkspaceConfigBase(BaseModel):
    """Base schema for workspace configuration."""
    allowed_channel_ids: List[str] = Field(default_factory=list)
    storage_mode: StorageModeEnum = StorageModeEnum.FULL_CONTENT
    default_agent_id: Optional[UUID] = None


class SlackWorkspaceConfigCreate(SlackWorkspaceConfigBase):
    """Schema for creating workspace configuration."""
    pass


class SlackWorkspaceConfigUpdate(BaseModel):
    """Schema for updating workspace configuration."""
    allowed_channel_ids: Optional[List[str]] = None
    storage_mode: Optional[StorageModeEnum] = None
    default_agent_id: Optional[UUID] = None


class SlackWorkspaceConfigResponse(SlackWorkspaceConfigBase):
    """Response schema for workspace configuration."""
    id: int
    organization_id: UUID
    team_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ==================== Agent Slack Config Schemas ====================

class AgentSlackConfigBase(BaseModel):
    """Base schema for agent Slack configuration."""
    channel_id: str
    channel_name: str
    enabled: bool = True
    respond_to_mentions: bool = True
    respond_to_reactions: bool = True
    respond_to_commands: bool = True
    reaction_emoji: str = "robot_face"


class AgentSlackConfigCreate(AgentSlackConfigBase):
    """Schema for creating agent Slack configuration."""
    pass


class AgentSlackConfigUpdate(BaseModel):
    """Schema for updating agent Slack configuration."""
    enabled: Optional[bool] = None
    respond_to_mentions: Optional[bool] = None
    respond_to_reactions: Optional[bool] = None
    respond_to_commands: Optional[bool] = None
    reaction_emoji: Optional[str] = None


class AgentSlackConfigResponse(AgentSlackConfigBase):
    """Response schema for agent Slack configuration."""
    id: int
    organization_id: UUID
    team_id: str
    agent_id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ==================== Bulk Configuration Schemas ====================

class AgentSlackConfigBulkCreate(BaseModel):
    """Schema for bulk creating/updating agent Slack configurations."""
    configs: List[AgentSlackConfigCreate]


class AgentSlackConfigBulkResponse(BaseModel):
    """Response schema for bulk agent Slack configuration."""
    agent_id: UUID
    configs: List[AgentSlackConfigResponse]


# ==================== OAuth State ====================

class SlackOAuthState(BaseModel):
    """Schema for OAuth state parameter."""
    organization_id: str
    nonce: str
    timestamp: int


# ==================== Webhook Event Schemas ====================

class SlackEventChallenge(BaseModel):
    """Schema for Slack URL verification challenge."""
    type: str
    token: str
    challenge: str


class SlackEventCallback(BaseModel):
    """Schema for Slack event callback."""
    type: str
    event: dict
    team_id: str
    api_app_id: str
    event_id: str
    event_time: int


class SlackSlashCommand(BaseModel):
    """Schema for Slack slash command."""
    command: str
    text: str
    response_url: str
    trigger_id: str
    user_id: str
    user_name: str
    channel_id: str
    channel_name: str
    team_id: str
    team_domain: str


class SlackInteractionPayload(BaseModel):
    """Schema for Slack interaction payload."""
    type: str
    callback_id: Optional[str] = None
    trigger_id: str
    user: dict
    channel: dict
    team: dict
    message: Optional[dict] = None
    response_url: str


# ==================== Data Deletion Schemas ====================

class SlackDataDeletionRequest(BaseModel):
    """Schema for Slack data deletion request."""
    team_id: str
    user_id: Optional[str] = None  # If None, delete all workspace data


class SlackDataDeletionResponse(BaseModel):
    """Response schema for data deletion."""
    ok: bool
    deleted_records: Optional[dict] = None
    message: Optional[str] = None
